File: backend/app/routers/drafts.py
import json
import sqlite3
from datetime import datetime
from pathlib import Path
from typing import Any
from urllib.parse import quote
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from ..db.database import get_db
from ..engines.applications.draft_generator import generate_draft_answers
from ..engines.applications.form_analyzer import analyze_form
from ..engines.applications.submission_engine import (
    BrowserUnavailableError,
    RateLimitError,
    confirm_submit_application,
    get_submission_progress,
    submit_application,
)

router = APIRouter(prefix="/drafts", tags=["drafts"])

JSON_COLUMNS = {"form_structure_json", "filled_answers_json", "screening_answers_json"}


class PrepareDraftRequest(BaseModel):
    job_id: str
    resume_version_id: str | None = None


class UpdateDraftRequest(BaseModel):
    filled_answers_json: dict[str, Any]


class AssistedFillRequest(BaseModel):
    confirm_user_assisted: bool = False
    acknowledge_platform_terms: bool = False
    use_visible_browser: bool = False
    pause_for_manual_input_seconds: int = 0


class AssistedFinalSubmitRequest(BaseModel):
    confirm_user_assisted: bool = False
    acknowledge_platform_terms: bool = False
    confirm_final_submit: bool = False
    use_visible_browser: bool = False


def db_conn():
    conn = get_db()
    try:
        yield conn
    finally:
        conn.close()


def _assert_assisted_consent(confirm_user_assisted: bool, acknowledge_platform_terms: bool) -> None:
    if not confirm_user_assisted:
        raise HTTPException(status_code=400, detail="User-assisted mode confirmation is required")
    if not acknowledge_platform_terms:
        raise HTTPException(status_code=400, detail="Platform terms acknowledgement is required")


def _row_to_draft(row: sqlite3.Row) -> dict[str, Any]:
    item = dict(row)
    for col in JSON_COLUMNS:
        raw = item.get(col)
        if isinstance(raw, str):
            try:
                item[col] = json.loads(raw)
            except json.JSONDecodeError:
                pass
    return item


def _artifact_url_for_screenshot_path(path: str | None) -> str | None:
    if not path:
        return None
    filename = Path(path).name.strip()
    if not filename:
        return None
    return f"/api/artifacts/screenshots/{quote(filename)}"


def _get_draft_or_404(draft_id: str, db: sqlite3.Connection) -> dict[str, Any]:
    row = db.execute("SELECT * FROM application_drafts WHERE id = ?", (draft_id,)).fetchone()
    if row is None:
        raise HTTPException(status_code=404, detail="Draft not found")
    return _row_to_draft(row)


@router.post("/prepare")
async def prepare_draft(payload: PrepareDraftRequest, db: sqlite3.Connection = Depends(db_conn)):
    job = db.execute(
        "SELECT * FROM jobs WHERE id = ? AND is_archived = 0",
        (payload.job_id,),
    ).fetchone()
    if job is None:
        raise HTTPException(status_code=404, detail="Job not found")

    profile = db.execute("SELECT * FROM user_profile WHERE id = 'local'").fetchone()
    if profile is None:
        raise HTTPException(status_code=404, detail="Profile not found")

    if payload.resume_version_id:
        resume = db.execute("SELECT id FROM resume_versions WHERE id = ?", (payload.resume_version_id,)).fetchone()
        if resume is None:
            raise HTTPException(status_code=404, detail="Resume version not found")

    job_dict = dict(job)
    profile_dict = dict(profile)

    form_fields = await analyze_form(str(job_dict.get("source_url") or ""))
    filled_answers = generate_draft_answers(job_dict, profile_dict, form_fields)

    draft_id = f"app-{uuid4().hex[:8]}"
    db.execute(
        """
        INSERT INTO application_drafts (
            id,
            job_id,
            resume_version_id,
            status,
            form_structure_json,
            filled_answers_json,
            created_at
        )
        VALUES (?, ?, ?, 'drafted', ?, ?, ?)
        """,
        (
            draft_id,
            payload.job_id,
            payload.resume_version_id,
            json.dumps(form_fields),
            json.dumps(filled_answers),
            datetime.utcnow().isoformat(),
        ),
    )
    db.commit()
    return _get_draft_or_404(draft_id, db)


@router.get("/{draft_id}")
def get_draft(draft_id: str, db: sqlite3.Connection = Depends(db_conn)):
    return _get_draft_or_404(draft_id, db)


@router.patch("/{draft_id}")
def update_draft(draft_id: str, payload: UpdateDraftRequest, db: sqlite3.Connection = Depends(db_conn)):
    result = db.execute(
        "UPDATE application_drafts SET filled_answers_json = ? WHERE id = ?",
        (json.dumps(payload.filled_answers_json), draft_id),
    )
    db.commit()
    if result.rowcount == 0:
        raise HTTPException(status_code=404, detail="Draft not found")
    return _get_draft_or_404(draft_id, db)


@router.post("/{draft_id}/approve")
def approve_draft(draft_id: str, db: sqlite3.Connection = Depends(db_conn)):
    result = db.execute(
        "UPDATE application_drafts SET status = 'approved', approved_at = ? WHERE id = ?",
        (datetime.utcnow().isoformat(), draft_id),
    )
    db.commit()
    if result.rowcount == 0:
        raise HTTPException(status_code=404, detail="Draft not found")
    return _get_draft_or_404(draft_id, db)


@router.post("/{draft_id}/reject")
def reject_draft(draft_id: str, db: sqlite3.Connection = Depends(db_conn)):
    result = db.execute(
        "UPDATE application_drafts SET status = 'rejected' WHERE id = ?",
        (draft_id,),
    )
    db.commit()
    if result.rowcount == 0:
        raise HTTPException(status_code=404, detail="Draft not found")
    return _get_draft_or_404(draft_id, db)


@router.post("/{draft_id}/submit")
async def submit_draft(
    draft_id: str,
    payload: AssistedFillRequest,
    db: sqlite3.Connection = Depends(db_conn),
):
    _assert_assisted_consent(payload.confirm_user_assisted, payload.acknowledge_platform_terms)
    try:
        result = await submit_application(
            draft_id,
            db,
            use_visible_browser=payload.use_visible_browser,
            pause_for_manual_input_seconds=payload.pause_for_manual_input_seconds,
        )
        screenshot_path = result.get("screenshot_path")
        return {
            "status": result.get("status", "ready_for_final_approval"),
            "screenshot_path": screenshot_path,
            "screenshot_url": _artifact_url_for_screenshot_path(screenshot_path),
            "mode": result.get("mode"),
            "requires_explicit_final_submit": True,
        }
    except RateLimitError as err:
        raise HTTPException(status_code=429, detail=str(err)) from err
    except BrowserUnavailableError as err:
        raise HTTPException(status_code=503, detail=str(err)) from err
    except ValueError as err:
        raise HTTPException(status_code=400, detail=str(err)) from err


@router.post("/{draft_id}/confirm-submit")
async def confirm_submit_draft(
    draft_id: str,
    payload: AssistedFinalSubmitRequest,
    db: sqlite3.Connection = Depends(db_conn),
):
    _assert_assisted_consent(payload.confirm_user_assisted, payload.acknowledge_platform_terms)
    if not payload.confirm_final_submit:
        raise HTTPException(status_code=400, detail="Final submit confirmation is required")
    try:
        result = await confirm_submit_application(
            draft_id,
            db,
            use_visible_browser=payload.use_visible_browser,
        )
    except RateLimitError as err:
        raise HTTPException(status_code=429, detail=str(err)) from err
    except BrowserUnavailableError as err:
        raise HTTPException(status_code=503, detail=str(err)) from err
    except ValueError as err:
        raise HTTPException(status_code=400, detail=str(err)) from err

    submitted_at = datetime.utcnow().isoformat()
    db.execute(
        """
        UPDATE application_drafts
        SET status = 'submitted', submitted_at = ?
        WHERE id = ?
        """,
        (submitted_at, draft_id),
    )

    run_id = f"submission-{uuid4().hex[:10]}"
    db.execute(
        """
        INSERT INTO discovery_runs (id, started_at, completed_at, jobs_found, jobs_new, source, status)
        VALUES (?, ?, ?, 0, 0, 'submission_engine', 'submitted')
        """,
        (run_id, submitted_at, submitted_at),
    )
    db.commit()

    draft = _get_draft_or_404(draft_id, db)
    screenshot_path = result.get("screenshot_path")
    return {
        "status": result.get("status", "submitted"),
        "screenshot_path": screenshot_path,
        "screenshot_url": _artifact_url_for_screenshot_path(screenshot_path),
        "mode": result.get("mode"),
        "draft": draft,
    }


@router.get("/{draft_id}/progress")
def get_draft_progress(
    draft_id: str,
    db: sqlite3.Connection = Depends(db_conn),
):
    _get_draft_or_404(draft_id, db)
    progress = get_submission_progress(draft_id)
    screenshot_path = progress.get("latest_screenshot_path")
    return {
        **progress,
        "latest_screenshot_url": _artifact_url_for_screenshot_path(
            screenshot_path if isinstance(screenshot_path, str) else None
        ),
    }

File: packages/api/src/drafts.ts
import type { ApiResponse } from './types';

export interface AssistedDraft {
  id: string;
  status: string;
}

export interface AssistedFillResult {
  status: string;
  screenshot_path?: string;
  screenshot_url?: string;
  mode?: string;
  requires_explicit_final_submit: boolean;
}

export interface AssistedConfirmResult {
  status: string;
  screenshot_path?: string;
  screenshot_url?: string;
  mode?: string;
  draft?: {
    id?: string;
    status?: string;
    submitted_at?: string | null;
  };
}

export interface AssistedProgressEvent {
  at: string;
  level: string;
  message: string;
}

export interface AssistedProgressResult {
  draft_id: string;
  status: string;
  mode: string;
  started_at?: string | null;
  updated_at?: string | null;
  latest_screenshot_path?: string;
  latest_screenshot_url?: string;
  events: AssistedProgressEvent[];
  error?: string | null;
}

export interface AssistedGuidanceResult {
  ok: boolean;
  draft_id: string;
  applied_guidance: string;
}

export interface AssistedRunOptions {
  useVisibleBrowser?: boolean;
  pauseForManualInputSeconds?: number;
}

function toErrorMessage(payload: unknown, fallback: string): string {
  if (payload && typeof payload === 'object' && 'detail' in payload) {
    const detail = (payload as { detail?: unknown }).detail;
    if (typeof detail === 'string' && detail.trim()) return detail;
  }
  return fallback;
}

export async function prepareDraft(jobId: string): Promise<ApiResponse<AssistedDraft>> {
  const response = await fetch('/api/drafts/prepare', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ job_id: jobId }),
  });
  const payload = await response.json();
  if (!response.ok) {
    return {
      data: { id: '', status: 'error' },
      error: toErrorMessage(payload, 'Failed to prepare draft'),
      status: response.status,
    };
  }
  return {
    data: {
      id: String(payload.id ?? ''),
      status: String(payload.status ?? 'drafted'),
    },
    status: response.status,
  };
}

export async function approveDraft(draftId: string): Promise<ApiResponse<AssistedDraft>> {
  const response = await fetch(`/api/drafts/${encodeURIComponent(draftId)}/approve`, {
    method: 'POST',
  });
  const payload = await response.json();
  if (!response.ok) {
    return {
      data: { id: draftId, status: 'error' },
      error: toErrorMessage(payload, 'Failed to approve draft'),
      status: response.status,
    };
  }
  return {
    data: {
      id: String(payload.id ?? draftId),
      status: String(payload.status ?? 'approved'),
    },
    status: response.status,
  };
}

export async function runAssistedFill(
  draftId: string,
  options?: AssistedRunOptions
): Promise<ApiResponse<AssistedFillResult>> {
  const response = await fetch(`/api/drafts/${encodeURIComponent(draftId)}/submit`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      confirm_user_assisted: true,
      acknowledge_platform_terms: true,
      use_visible_browser: Boolean(options?.useVisibleBrowser),
      pause_for_manual_input_seconds: Number(options?.pauseForManualInputSeconds ?? 0),
    }),
  });
  const payload = await response.json();
  if (!response.ok) {
    return {
      data: {
        status: 'error',
        requires_explicit_final_submit: true,
      },
      error: toErrorMessage(payload, 'Failed during assisted form fill'),
      status: response.status,
    };
  }
  return {
    data: {
      status: String(payload.status ?? 'ready_for_final_approval'),
      screenshot_path:
        typeof payload.screenshot_path === 'string' ? payload.screenshot_path : undefined,
      screenshot_url:
        typeof payload.screenshot_url === 'string' ? payload.screenshot_url : undefined,
      mode: typeof payload.mode === 'string' ? payload.mode : undefined,
      requires_explicit_final_submit: Boolean(payload.requires_explicit_final_submit ?? true),
    },
    status: response.status,
  };
}

export async function runAssistedConfirmSubmit(
  draftId: string,
  options?: AssistedRunOptions
): Promise<ApiResponse<AssistedConfirmResult>> {
  const response = await fetch(`/api/drafts/${encodeURIComponent(draftId)}/confirm-submit`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      confirm_user_assisted: true,
      acknowledge_platform_terms: true,
      confirm_final_submit: true,
      use_visible_browser: Boolean(options?.useVisibleBrowser),
    }),
  });
  const payload = await response.json();
  if (!response.ok) {
    return {
      data: { status: 'error' },
      error: toErrorMessage(payload, 'Failed during final submit'),
      status: response.status,
    };
  }
  const draftPayload =
    payload.draft && typeof payload.draft === 'object'
      ? (payload.draft as { id?: unknown; status?: unknown; submitted_at?: unknown })
      : undefined;
  return {
    data: {
      status: String(payload.status ?? 'submitted'),
      screenshot_path:
        typeof payload.screenshot_path === 'string' ? payload.screenshot_path : undefined,
      screenshot_url:
        typeof payload.screenshot_url === 'string' ? payload.screenshot_url : undefined,
      mode: typeof payload.mode === 'string' ? payload.mode : undefined,
      draft: draftPayload
        ? {
            id: typeof draftPayload.id === 'string' ? draftPayload.id : undefined,
            status: typeof draftPayload.status === 'string' ? draftPayload.status : undefined,
            submitted_at:
              typeof draftPayload.submitted_at === 'string' ? draftPayload.submitted_at : null,
          }
        : undefined,
    },
    status: response.status,
  };
}

export async function getAssistedProgress(
  draftId: string
): Promise<ApiResponse<AssistedProgressResult>> {
  const response = await fetch(`/api/drafts/${encodeURIComponent(draftId)}/progress`);
  const payload = await response.json();
  if (!response.ok) {
    return {
      data: {
        draft_id: draftId,
        status: 'error',
        mode: 'unknown',
        events: [],
      },
      error: toErrorMessage(payload, 'Failed to fetch assisted progress'),
      status: response.status,
    };
  }
  const rawEvents: unknown[] = Array.isArray(payload.events) ? payload.events : [];
  const events = rawEvents
    .map((item: unknown) =>
      item && typeof item === 'object'
        ? (() => {
            const event = item as Record<string, unknown>;
            return {
              at: typeof event.at === 'string' ? event.at : '',
              level: typeof event.level === 'string' ? event.level : 'info',
              message: typeof event.message === 'string' ? event.message : '',
            };
          })()
        : null
    )
    .filter((item: AssistedProgressEvent | null): item is AssistedProgressEvent =>
      Boolean(item && item.message)
    );

  return {
    data: {
      draft_id: typeof payload.draft_id === 'string' ? payload.draft_id : draftId,
      status: typeof payload.status === 'string' ? payload.status : 'idle',
      mode: typeof payload.mode === 'string' ? payload.mode : 'unknown',
      started_at: typeof payload.started_at === 'string' ? payload.started_at : null,
      updated_at: typeof payload.updated_at === 'string' ? payload.updated_at : null,
      latest_screenshot_path:
        typeof payload.latest_screenshot_path === 'string' ? payload.latest_screenshot_path : undefined,
      latest_screenshot_url:
        typeof payload.latest_screenshot_url === 'string' ? payload.latest_screenshot_url : undefined,
      events,
      error: typeof payload.error === 'string' ? payload.error : null,
    },
    status: response.status,
  };
}

export async function sendAssistedGuidance(
  draftId: string,
  message: string
): Promise<ApiResponse<AssistedGuidanceResult>> {
  const response = await fetch(`/api/drafts/${encodeURIComponent(draftId)}/guidance`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ message }),
  });
  const payload = await response.json();
  if (!response.ok) {
    return {
      data: { ok: false, draft_id: draftId, applied_guidance: '' },
      error: toErrorMessage(payload, 'Failed to send operator guidance'),
      status: response.status,
    };
  }
  return {
    data: {
      ok: Boolean(payload.ok),
      draft_id: typeof payload.draft_id === 'string' ? payload.draft_id : draftId,
      applied_guidance:
        typeof payload.applied_guidance === 'string' ? payload.applied_guidance : '',
    },
    status: response.status,
  };
}
